src/OptimalControlUtilities/include/BipedalLocomotionControllers/OptimalControlUtilities/OptimizationProblemElements.h

```diff
@@ -0,0 +1,254 @@
+
+    /**
+     * Get the inequality constraints
+     * @return a vector containing the reference to the inequality constraints
+     */
+    const std::vector<InequalityConstraint>& getInequalityConstraints() const;
+};
+
+/**
+ * CostFunction class handles the cost function for a QP
+ * optimization problem. i.e. \f$\frac{1}{2} (x ^ \top H x + g^\top x)\f$ where
+ * \f$H\f$ is the hessian matrix
+ * \f$g\f$ is the gradient vector
+ */
+class CostFunction
+{
+    /**
+     * CostFunctionElement contains an element of the cost function
+     */
+    struct CostFunctionElement
+    {
+        ControlTask* element; /**< Pointer to the element */
+        iDynTree::VectorDynSize weight; /**< Weight */
+        double weightOffset{0}; /**< Offset of the weight i.e. Weight = weightScaling * customWeight
+                                   + weightOffset */
+        double weightScaling{1}; /**< Scaling factor of the weight i.e. Weight = weightScaling *
+                                    customWeight + weightOffset */
+
+        /**
+         * Constructor
+         */
+        CostFunctionElement(ControlTask* const element,
+                            const iDynTree::VectorDynSize& weight,
+                            const double& weightScaling = 1,
+                            const double& weightOffset = 0);
+    };
+
+    /** Unordered map containing all the cost function element */
+    std::unordered_map<std::string, CostFunctionElement> m_costFunctionElements;
+
+    iDynTree::VectorDynSize m_gradient; /**< Gradient vector */
+    iDynTree::MatrixDynSize m_hessianMatrix; /**< Hessian matrix */
+
+public:
+    /**
+     * Elements wraps an std::pair<iDynTree::MatrixDynSize&, iDynTree::VectorDynSize&>. It contains
+     * the hessian matrix and the gradient vector
+     */
+    class Elements : public std::pair<iDynTree::MatrixDynSize&, iDynTree::VectorDynSize&>
+    {
+        // obscure first and second attribute
+        using std::pair<iDynTree::MatrixDynSize&, iDynTree::VectorDynSize&>::first;
+        using std::pair<iDynTree::MatrixDynSize&, iDynTree::VectorDynSize&>::second;
+
+    public:
+        // use the std::pair constructors
+        using std::pair<iDynTree::MatrixDynSize&, iDynTree::VectorDynSize&>::pair;
+
+        iDynTree::MatrixDynSize& hessian();
+        const iDynTree::MatrixDynSize& hessian() const;
+        iDynTree::VectorDynSize& gradient();
+        const iDynTree::VectorDynSize& gradient() const;
+    };
+
+    /**
+     * Constructor
+     * @param handler variable handler object
+     */
+    CostFunction(const VariableHandler& handler);
+
+    /**
+     * Add a new cost function element.
+     * @param element pointer to the cost function element
+     * @param weight cost weight
+     * @param weightScaling Scaling factor of the weight
+     * @param weightOffset offset of the weight
+     * @param name name associated to the element
+     */
+    void addCostFunction(ControlTask* element,
```

Comment:
``addCostElement``? ``addCostTerm``? The cost function is the sum of all the terms.